signature/src/mechanism/cosign/mod.rs

```diff
@@ -0,0 +1,346 @@
+    //
+    // This field is optional.
+    #[serde(default, rename = "signedIdentity")]
+    pub signed_identity: Option<PolicyReqMatchType>,
+}
+
+#[async_trait]
+impl SignScheme for CosignParameters {
+    /// This initialization will:
+    /// * Create [`COSIGN_KEY_DIR`] if not exist.
+    async fn init(&self) -> Result<()> {
+        if !Path::new(COSIGN_KEY_DIR).exists() {
+            fs::create_dir_all(COSIGN_KEY_DIR).await.map_err(|e| {
+                anyhow!("Create Simple Signing sigstore-config dir failed: {:?}", e)
+            })?;
+        }
+
+        Ok(())
+    }
+
+    fn resource_manifest(&self) -> HashMap<&str, &str> {
+        let mut manifest = HashMap::new();
+        if let Some(key_path) = &self.key_path {
+            if !Path::new(key_path).exists() {
+                manifest.insert(COSIGN_KEY_KBS, &key_path[..]);
+            }
+        }
+        manifest
+    }
+
+    /// Judge whether an image is allowed by this SignScheme.
+    async fn allows_image(&self, image: &mut Image) -> Result<()> {
+        // Check before we access the network
+        self.check_reference_rule_types()?;
+
+        // Verification, will access the network
+        let payloads = self.verify_signature_and_get_payload(image).await?;
+
+        // check the reference rules (signed identity)
+        for payload in payloads {
+            if let Some(rule) = &self.signed_identity {
+                payload.validate_signed_docker_reference(&image.reference, rule)?;
+            }
+
+            payload.validate_signed_docker_manifest_digest(&image.manifest_digest.to_string())?;
+        }
+
+        Ok(())
+    }
+}
+
+impl CosignParameters {
+    /// Check whether this Policy Request Match Type (i.e., signed identity
+    /// check type) for the reference
+    /// is MatchRepository or ExactRepository.
+    /// Because cosign-created signatures only contain a repository,
+    /// so only matchRepository and exactRepository can be used to accept them.
+    /// Other types are all to be denied.
+    /// If it is neither of them, return `Error`. Otherwise, return `Ok()`
+    fn check_reference_rule_types(&self) -> Result<()> {
+        match &self.signed_identity {
+            Some(rule) => match rule {
+                PolicyReqMatchType::MatchRepository
+                | PolicyReqMatchType::ExactRepository { .. } => Ok(()),
+                p => Err(anyhow!("Denied by {:?}", p)),
+            },
+            None => Ok(()),
+        }
+    }
+
+    /// Verify the cosign-signed image. There will be three steps:
+    /// * Get the pub key.
+    /// * Download the cosign-signed image's manifest and its digest. Calculate its
+    /// signature's image.
+    /// * Download the signature image, gather the signatures and verify them
+    /// using the pubkey.
+    /// If succeeds, the payloads of the signature will be returned.
+    async fn verify_signature_and_get_payload(&self, image: &mut Image) -> Result<Vec<SigPayload>> {
```

Comment:
Could you add a unit test for this method?

Reply:
Added